plugins/inputs/turbostat/turbostat.go

```diff
@@ -0,0 +1,226 @@
+
+	command []string
+	process *process.Process
+}
+
+type column struct {
+	name   string
+	isTag  bool
+	isTime bool
+}
+
+func (*Turbostat) SampleConfig() string {
+	return sampleConfig
+}
+
+// Init initialises the plugin.
+func (t *Turbostat) Init() error {
+	if t.Path == "" {
+		t.Path = "turbostat"
+	}
+
+	if t.Interval <= 0 {
+		t.Interval = config.Duration(10 * time.Second)
+	}
+
+	// Build the command line to spawn Turbostat.
+	t.command = make([]string, 0, 10)
+	if t.UseSudo {
+		t.command = append(t.command, "sudo")
+	}
+	s := int(time.Duration(t.Interval).Seconds())
+	t.command = append(t.command, t.Path, "--quiet", "--interval", strconv.Itoa(s), "--show", "all")
+
+	return nil
+}
+
+// Start starts the plugin.
+func (t *Turbostat) Start(acc telegraf.Accumulator) error {
+	var err error
+	t.process, err = process.New(t.command, nil)
+	if err != nil {
+		return fmt.Errorf("failed to create process %s: %w", t.command, err)
+	}
+
+	// Read error lines from Turbostat stderr and add them to the accumulator.
+	t.process.ReadStderrFn = func(r io.Reader) {
+		scanner := bufio.NewScanner(r)
+		for scanner.Scan() {
+			acc.AddError(errors.New(scanner.Text()))
+		}
+		acc.AddError(scanner.Err())
+	}
+
+	// Read metrics from Turbostat stdout and add them to the accumulator.
+	// Add any error to the accumulator and continue processing the next lines.
+	t.process.ReadStdoutFn = func(r io.Reader) {
+		scanner := bufio.NewScanner(r)
+		if !scanner.Scan() {
+			acc.AddError(scanner.Err())
+			return
+		}
+		header := scanner.Text()
+		columns := make([]column, 0, 50)
+		for _, s := range strings.Fields(header) {
+			columns = append(columns, createColumn(s))
+		}
+		for scanner.Scan() {
+			line := scanner.Text()
+			if line == header {
+				continue
+			}
+			values := strings.Fields(line)
+			acc.AddError(processValues(acc, columns, values))
+		}
+		acc.AddError(scanner.Err())
+	}
+
+	t.process.StopOnError = false
+	t.process.Log = t.Log
+
+	if err = t.process.Start(); err != nil {
+		return fmt.Errorf("failed to start process %s: %w", t.command, err)
+	}
+	return nil
+}
+
+// Stop stops the plugin.
+func (t *Turbostat) Stop() {
+	t.process.Stop()
+}
+
+func (*Turbostat) Gather(telegraf.Accumulator) error {
+	return nil
+}
+
+// Creates a metric from columns and values, and adds it to an accumulator.
+func processValues(acc telegraf.Accumulator, columns []column, values []string) error {
+	if len(values) > len(columns) {
+		return fmt.Errorf("too many values: %d columns, %d values", len(columns), len(values))
+	}
+	tags := make(map[string]string, len(values))
+	fields := make(map[string]any, len(values))
+	timestamp := time.Now()
+	for i, value := range values {
+		column := columns[i]
+		switch {
+		case column.isTime:
+			f, err := strconv.ParseFloat(value, 64)
+			if err != nil {
+				return fmt.Errorf("unable to parse time: %w", err)
+			}
+			s, ns := math.Modf(f)
+			timestamp = time.Unix(int64(s), int64(ns*(1e9)))
```

Comment:
This should be replaced by

```suggestion
			timestamp = time.Unix(0, int64(f*1e9))
```

which does not require the `math` function and works.

Reply:
Nice, thanks.